File: src/rs/cache/CacheIndex.ts
import { CompressionHandler } from "../compression/CompressionHandler";
import { XteaKey } from "../crypto/Xtea";
import { ByteSource } from "../io/ByteSource";
import { ByteSourceReader } from "../io/ByteSourceReader";
import { getOrCopyBytes, readAllBytes } from "../io/ByteSourceUtil";
import { Uint8ArrayByteSource } from "../io/Uint8ArrayByteSource";
import { StringUtil } from "../util/StringUtil";
import { DatIndexId } from "./IndexId";
import { Archive } from "./format/Archive";
import { ArchiveFile } from "./format/ArchiveFile";
import { Container } from "./format/Container";
import { ArchiveReference } from "./reference/ArchiveReference";
import { ReferenceTable } from "./reference/ReferenceTable";
import { CacheStore } from "./store/CacheStore";
import { IDX_ENTRY_SIZE } from "./store/DatLayout";

export abstract class CacheIndex {
    static readonly META_INDEX_ID: i32 = 255;

    constructor(
        readonly id: number,
        readonly compressionHandler: CompressionHandler,
    ) {}

    // Only store-backed indices (Dat/Dat2) expose raw on-disk archive bytes.
    readonly store?: CacheStore;

    abstract getArchiveIds(): Int32Array;

    abstract getArchiveCount(): number;

    abstract getLastArchiveId(): number;

    abstract getArchiveReference(archiveId: number): ArchiveReference | null;

    abstract getArchiveId(name: string): number;

    tryGetArchiveId(name: string): number | undefined {
        const id = this.getArchiveId(name);
        return id === -1 ? undefined : id;
    }

    getFileIds(archiveId: number): Int32Array | null {
        const ref = this.getArchiveReference(archiveId);
        return ref ? ref.fileIds : null;
    }

    abstract archiveExists(archiveId: number): boolean;

    abstract getFileCount(archiveId: number): number;

    abstract getArchiveKey(archiveId: number, key: XteaKey | null): Archive;

    tryGetArchiveKey(archiveId: number, key: XteaKey | null): Archive | undefined {
        try {
            return this.getArchiveKey(archiveId, key);
        } catch {
            return undefined;
        }
    }

    /**
     * Reads the raw on-disk bytes for this archive (container/packed format as stored in the cache).
     *
     * Note: only available for store-backed indices (Dat/Dat2). Legacy/Classic indices are decoded
     * up-front and do not retain raw archive bytes.
     */
    readArchiveBytes(archiveId: number): Uint8Array {
        const store = this.store;
        if (!store) {
            throw new Error("readArchiveBytes() unsupported (no store)");
        }
        let rawSource: ByteSource;
        try {
            rawSource = store.openArchiveReader(this.id, archiveId);
        } catch {
            return new Uint8Array(0);
        }
        if (rawSource.size === 0) {
            return new Uint8Array(0);
        }
        try {
            return readAllBytes(rawSource);
        } catch {
            return new Uint8Array(0);
        }
    }

    /**
     * Dat2-only: decodes the container and returns the payload bytes (archive format bytes).
     */
    readContainerPayload(_archiveId: number, _key: XteaKey | null): Uint8Array {
        throw new Error("readContainerPayload() unsupported");
    }

    tryReadContainerPayload(archiveId: number, key: XteaKey | null): Uint8Array | undefined {
        try {
            return this.readContainerPayload(archiveId, key);
        } catch {
            return undefined;
        }
    }

    getArchive(archiveId: number): Archive {
        return this.getArchiveKey(archiveId, null);
    }

    tryGetArchive(archiveId: number): Archive | undefined {
        return this.tryGetArchiveKey(archiveId, null);
    }

    getFileKey(archiveId: number, fileId: number, key: XteaKey | null): ArchiveFile | null {
        return this.getArchiveKey(archiveId, key).getFile(fileId);
    }

    getFileSmart(id: number, key: XteaKey | null): ArchiveFile | null {
        if (this.getArchiveCount() === 1) {
            return this.getFileKey(0, id, key);
        } else if (this.getFileCount(id) === 1) {
            return this.getFileKey(id, 0, key);
        }
        throw new Error("Invalid archive");
    }

    tryGetFileKey(archiveId: number, fileId: number, key: XteaKey | null): ArchiveFile | undefined {
        const archive = this.tryGetArchiveKey(archiveId, key);
        return archive?.getFile(fileId) ?? undefined;
    }

    tryGetFileSmart(id: number, key: XteaKey | null): ArchiveFile | undefined {
        if (this.getArchiveCount() === 1) {
            return this.tryGetFileKey(0, id, key);
        } else if (this.getFileCount(id) === 1) {
            return this.tryGetFileKey(id, 0, key);
        }
        return undefined;
    }

    getFile(archiveId: number, fileId: number): ArchiveFile | null {
        return this.getFileKey(archiveId, fileId, null);
    }

    tryGetFile(archiveId: number, fileId: number): ArchiveFile | undefined {
        return this.tryGetFileKey(archiveId, fileId, null);
    }
}

function byteSourceFromBytes(data: Uint8Array): ByteSource {
    return new Uint8ArrayByteSource(data);
}

function decodeTableFromSource(
    source: ByteSource,
    compressionHandler: CompressionHandler,
): ReferenceTable {
    if (source.size === 0) {
        return ReferenceTable.INVALID_TABLE;
    }
    const container = Container.decodeFromSource(source, null, compressionHandler);
    return ReferenceTable.decodeFromReader(
        new ByteSourceReader(byteSourceFromBytes(container.data)),
    );
}

function decodeArchiveDataFromSource(
    table: ReferenceTable,
    compressionHandler: CompressionHandler,
    archiveId: number,
    source: ByteSource,
    key: XteaKey | null,
): Archive {
    const archiveRef = table.getArchiveReference(archiveId);
    if (!archiveRef) {
        throw new Error("Archive reference not found for: " + archiveId);
    }
    const container = Container.decodeFromSource(source, key, compressionHandler);
    return Archive.decodeFromSource(archiveRef, byteSourceFromBytes(container.data));
}

export class DatCacheIndex extends CacheIndex {
    private _archiveIds: Int32Array | null = null;

    private constructor(
        id: number,
        private readonly _archiveCount: number,
        readonly store: CacheStore,
        compressionHandler: CompressionHandler,
    ) {
        super(id, compressionHandler);
    }

    static fromStore(
        id: number,
        store: CacheStore,
        compressionHandler: CompressionHandler,
    ): DatCacheIndex {
        const indexSize = store.getIndexFileSize(id);
        if (indexSize === null) {
            throw new Error("Index file not found: " + id);
        }
        if (indexSize % IDX_ENTRY_SIZE !== 0) {
            throw new Error(`Invalid .idx size: ${indexSize} (not divisible by ${IDX_ENTRY_SIZE})`);
        }
        const archiveCount = indexSize / IDX_ENTRY_SIZE;
        return new DatCacheIndex(id, archiveCount, store, compressionHandler);
    }

    getArchiveIds(): Int32Array {
        if (this._archiveIds) {
            return this._archiveIds;
        }
        const archiveIds = new Int32Array(this._archiveCount);
        for (let i = 0; i < this._archiveCount; i++) {
            archiveIds[i] = i;
        }
        this._archiveIds = archiveIds;
        return archiveIds;
    }

    getArchiveCount(): number {
        return this._archiveCount;
    }

    getLastArchiveId(): number {
        return this._archiveCount - 1;
    }

    getArchiveReference(_archiveId: number): ArchiveReference | null {
        return null;
    }

    getArchiveId(_name: string): number {
        return -1;
    }

    archiveExists(archiveId: number): boolean {
        return archiveId >= 0 && archiveId < this._archiveCount;
    }

    getFileCount(_archiveId: number): number {
        return 0;
    }

    getArchiveKey(archiveId: number, _key: XteaKey | null): Archive {
        if (!this.archiveExists(archiveId)) {
            throw new Error("Archive not found: " + archiveId);
        }
        const data = getOrCopyBytes(this.store.openArchiveReader(this.id, archiveId));
        return Archive.decodeOld(
            archiveId,
            data,
            this.id === DatIndexId.configs,
            this.compressionHandler,
        );
    }
}

export class Dat2CacheIndex extends CacheIndex {
    private constructor(
        id: number,
        table: ReferenceTable,
        readonly store: CacheStore,
        compressionHandler: CompressionHandler,
    ) {
        super(id, compressionHandler);
        this.table = table;
    }

    readonly table: ReferenceTable;

    static fromDat2Store(
        id: number,
        store: CacheStore,
        compressionHandler: CompressionHandler,
    ): Dat2CacheIndex {
        const metaSource = store.openArchiveReader(CacheIndex.META_INDEX_ID, id);
        const table = decodeTableFromSource(metaSource, compressionHandler);
        return new Dat2CacheIndex(id, table, store, compressionHandler);
    }

    getArchiveIds(): Int32Array {
        return this.table.archiveIds;
    }

    getArchiveCount(): number {
        return this.table.archiveCount;
    }

    getLastArchiveId(): number {
        return this.table.lastArchiveId;
    }

    getArchiveReference(archiveId: number): ArchiveReference | null {
        return this.table.getArchiveReference(archiveId);
    }

    getArchiveId(name: string): number {
        const value = this.table.getArchiveId(name);
        return value ?? -1;
    }

    archiveExists(archiveId: number): boolean {
        return this.table.archiveExists(archiveId);
    }

    getFileCount(archiveId: number): number {
        const value = this.table.getArchiveReference(archiveId);
        return value ? value.fileCount : 0;
    }

    override getArchiveKey(archiveId: number, key: XteaKey | null): Archive {
        const source = this.store.openArchiveReader(this.id, archiveId);
        return decodeArchiveDataFromSource(
            this.table,
            this.compressionHandler,
            archiveId,
            source,
            key,
        );
    }

    override readContainerPayload(archiveId: number, key: XteaKey | null): Uint8Array {
        const raw = this.readArchiveBytes(archiveId);
        if (raw.byteLength === 0) {
            return new Uint8Array(0);
        }
        const container = Container.decodeFromSource(
            new Uint8ArrayByteSource(raw),
            key,
            this.compressionHandler,
        );
        return container.data;
    }
}

export class LegacyCacheIndex extends CacheIndex {
    private _archiveIds: Int32Array | null = null;

    constructor(
        readonly id: number,
        readonly archives: Archive[],
        compressionHandler: CompressionHandler,
        readonly archiveNameHashes: Map<number, number> = new Map(),
    ) {
        super(id, compressionHandler);
    }

    getArchiveIds(): Int32Array {
        if (this._archiveIds) {
            return this._archiveIds;
        }
        const archiveIds = new Int32Array(this.archives.length);
        for (let i = 0; i < this.archives.length; i++) {
            archiveIds[i] = i;
        }
        this._archiveIds = archiveIds;
        return archiveIds;
    }

    getArchiveCount(): number {
        return this.archives.length;
    }

    getLastArchiveId(): number {
        return this.archives.length - 1;
    }

    getArchiveReference(_archiveId: number): ArchiveReference | null {
        return null;
    }

    override getArchiveId(name: string): number {
        const value = this.archiveNameHashes.get(StringUtil.hashOld(name));
        return value ?? -1;
    }

    archiveExists(archiveId: number): boolean {
        return (
            archiveId >= 0 &&
            archiveId < this.archives.length &&
            this.archives[archiveId] !== undefined
        );
    }

    getFileCount(archiveId: number): number {
        const archive = this.archives[archiveId];
        return archive ? archive.fileCount : 0;
    }

    override getArchiveKey(archiveId: number, key: XteaKey | null): Archive {
        return this.archives[archiveId];
    }

    override getFileKey(
        archiveId: number,
        fileId: number,
        key: XteaKey | null,
    ): ArchiveFile | null {
        const value = this.archives[archiveId];
        return value ? value.getFile(fileId) : null;
    }
}

File: cpp/rs/cache/CacheIndex.cpp
#include "CacheIndex.hpp"

#include <cstddef>

#include "../compression/CompressionHandler.hpp"
#include "../core/Allocator.hpp"
#include "../core/Move.hpp"
#include "../core/Result.hpp"
#include "../core/Span.hpp"
#include "../core/Status.hpp"
#include "../core/Vec.hpp"
#include "../crypto/Xtea.hpp"
#include "../types.hpp"
#include "CacheType.hpp"
#include "format/Archive.hpp"
#include "../cache/ArchiveMeta.hpp"
#include "../cache/format/Container.hpp"
#include "reference/ReferenceTable.hpp"
#include "store/CacheStore.hpp"
#include "../cache/store/DatLayout.hpp"
#include "../io/ByteSourceReader.hpp"
#include "../io/Uint8ArrayByteSource.hpp"

namespace rs {

CacheIndex::CacheIndex(
    CacheType cacheType,
    i32 id,
    const CacheStore& store,
    const CompressionHandler& compressionHandler,
    i32 archiveCount,
    ReferenceTable table,
    Vec<i32> archiveIds) noexcept
    : cacheType_(cacheType),
      id_(id),
      store_(&store),
      compressionHandler_(&compressionHandler),
      archiveCount_(archiveCount),
      archiveIds_(rs::move(archiveIds)),
      table_(rs::move(table)) {
}

Result<CacheIndex> CacheIndex::fromStore(
    CacheType cacheType,
    i32 id,
    const CacheStore& store,
    const CompressionHandler& compressionHandler,
    Allocator& alloc) noexcept {
    if (cacheType != CacheType::Dat && cacheType != CacheType::Dat2) {
        return Result<CacheIndex>::err(Status::Unsupported);
    }

    if (cacheType == CacheType::Dat) {
        std::size_t indexSize = 0;
        const Status s = store.getIndexFileSize(id, &indexSize);
        if (!ok(s)) {
            return Result<CacheIndex>::err(s);
        }
        if ((indexSize % IDX_ENTRY_SIZE) != 0) {
            return Result<CacheIndex>::err(Status::BadFormat);
        }
        const i32 archiveCount = static_cast<i32>(indexSize / IDX_ENTRY_SIZE);
        Vec<i32> ids(alloc);
        ReferenceTable emptyTable;
        return Result<CacheIndex>::ok(CacheIndex(cacheType, id, store, compressionHandler, archiveCount, rs::move(emptyTable), rs::move(ids)));
    }

    // Dat2: meta index 255 contains a reference-table container per index id.
    Vec<u8> metaBytes(alloc);
    const Status s = store.readArchive(META_INDEX_ID, id, &metaBytes);
    if (!ok(s)) {
        return Result<CacheIndex>::err(s);
    }
    if (metaBytes.size() == 0) {
        // Mirror TS behavior: empty reference-table source => "invalid" (empty) table.
        Vec<i32> ids(alloc);
        ReferenceTable emptyTable;
        return Result<CacheIndex>::ok(CacheIndex(cacheType, id, store, compressionHandler, 0, rs::move(emptyTable), rs::move(ids)));
    }

    Uint8ArrayByteSource metaSource(metaBytes.data(), metaBytes.size());
    auto containerRes = Container::decodeFromSource(metaSource, nullptr, compressionHandler, alloc);
    if (!containerRes.isOk()) {
        return Result<CacheIndex>::err(containerRes.status());
    }
    Container container = rs::move(containerRes.value());

    Uint8ArrayByteSource tableSource(container.data.data(), container.data.size());
    ByteSourceReader reader(&tableSource);
    auto tableRes = ReferenceTable::decodeFromReader(reader, alloc);
    if (!tableRes.isOk()) {
        return Result<CacheIndex>::err(tableRes.status());
    }

    Vec<i32> ids(alloc);
    return Result<CacheIndex>::ok(CacheIndex(cacheType, id, store, compressionHandler, 0, rs::move(tableRes.value()), rs::move(ids)));
}

Span<const i32> CacheIndex::archiveIds() const noexcept {
    if (cacheType_ == CacheType::Dat2) {
        return table_.archiveIds();
    }
    if (!archiveIdsBuilt_) {
        archiveIds_.clear();
        auto r = archiveIds_.reserve(static_cast<std::size_t>(archiveCount_));
        if (r.isOk()) {
            for (i32 i = 0; i < archiveCount_; i++) {
                (void)archiveIds_.pushBack(i);
            }
        }
        archiveIdsBuilt_ = true;
    }
    return Span<const i32>(archiveIds_.data(), archiveIds_.size());
}

i32 CacheIndex::archiveCount() const noexcept {
    if (cacheType_ == CacheType::Dat2) {
        return table_.archiveCount();
    }
    return archiveCount_;
}

i32 CacheIndex::lastArchiveId() const noexcept {
    if (cacheType_ == CacheType::Dat2) {
        return table_.lastArchiveId();
    }
    return archiveCount_ > 0 ? (archiveCount_ - 1) : -1;
}

bool CacheIndex::archiveExists(i32 archiveId) const noexcept {
    if (cacheType_ == CacheType::Dat2) {
        return table_.archiveExists(archiveId);
    }
    return archiveId >= 0 && archiveId < archiveCount_;
}

i32 CacheIndex::fileCount(i32 archiveId) const noexcept {
    if (cacheType_ != CacheType::Dat2) {
        return 0;
    }
    ArchiveMeta meta;
    const Status s = table_.getArchiveMeta(archiveId, &meta);
    if (!ok(s)) {
        return 0;
    }
    return meta.fileCount;
}

Status CacheIndex::getArchiveMeta(i32 archiveId, ArchiveMeta* out) const noexcept {
    if (cacheType_ != CacheType::Dat2) {
        return Status::Unsupported;
    }
    return table_.getArchiveMeta(archiveId, out);
}

Result<Archive> CacheIndex::getArchiveKey(i32 archiveId, const XteaKey* key, Allocator& alloc) const noexcept {
    if (!store_ || !compressionHandler_) {
        return Result<Archive>::err(Status::InvalidArgument);
    }
    if (!archiveExists(archiveId)) {
        return Result<Archive>::err(Status::NotFound);
    }

    Vec<u8> bytes(alloc);
    const Status s = store_->readArchive(id_, archiveId, &bytes);
    if (!ok(s)) {
        return Result<Archive>::err(s);
    }

    if (cacheType_ == CacheType::Dat) {
        // Dat caches store "old" archive bytes directly.
        // Index 0 ("configs") is multi-file; other indices are typically single-file.
        const bool multipleFiles = (id_ == 0);
        return Archive::decodeOld(
            archiveId,
            Span<const u8>(bytes.data(), bytes.size()),
            multipleFiles,
            *compressionHandler_,
            alloc);
    }

    ArchiveMeta meta;
    const Status m = table_.getArchiveMeta(archiveId, &meta);
    if (!ok(m)) {
        return Result<Archive>::err(m);
    }

    Uint8ArrayByteSource source(bytes.data(), bytes.size());
    auto containerRes = Container::decodeFromSource(source, key, *compressionHandler_, alloc);
    if (!containerRes.isOk()) {
        return Result<Archive>::err(containerRes.status());
    }
    Container container = rs::move(containerRes.value());

    return Archive::decodeFromBytes(meta, Span<const u8>(container.data.data(), container.data.size()), alloc);
}

} // namespace rs
